roles/openshift_logging_elasticsearch/templates/2.x/es.j2

```diff
@@ -27,6 +27,32 @@ spec:
         component: "{{component}}"
         deployment: "{{deploy_name}}"
     spec:
+      affinity:
+        podAntiAffinity:
+          requiredDuringSchedulingIgnoredDuringExecution:
+          - labelSelector:
+              matchExpressions:
+              - key: logging-infra
+                operator: In
+                values:
+                - elasticsearch
+            topologyKey: kubernetes.io/hostname
+{% if openshift_infra_anti_affinity_labels is defined and openshift_infra_anti_affinity_labels is iterable and openshift_infra_anti_affinity_labels | length > 0 %}
+          preferredDuringSchedulingIgnoredDuringExecution:
```

Comment:
Sure, so when it comes to other elasticsearch pods, we want to avoid being scheduled on the same node (for the `required`). 

For `preferred` we want to try to balance out the distribution of any infra pods to minimize overloading a node. The biggest one would be if Cassandra is also deployed, however it shouldn't prevent us from scheduling all together.

Reply:
> Sure, so when it comes to other elasticsearch pods, we want to avoid being scheduled on the same node (for the required).

I agree with the above part. But are you sure always that number of instances is always less than/equal to number of nodes? otherwise extra instances will go into pending. 

> For preferred we want to try to balance out the distribution of any infra pods to minimize overloading a node. The biggest one would be if Cassandra is also deployed, however it shouldn't prevent us from scheduling all together.

I am more concerned about this part and I need more information so that I am sure it really does what you are trying to do here, so that we dont end up adding something that is not required or does not help.  

So when you say `any infra pods`, what it seems like that basically you are trying to have one such infra pod per node, even though its a soft requirement (preferredDuringSchedulingIgnoredDuringExecution) but unless number of pods is close enough to number of nodes, it wont be of much help here.

Reply:
> I agree with the above part. But are you sure always that number of instances is always less than/equal to number of nodes? otherwise extra instances will go into pending.

Yes, it has been pointed out by @portante that with more than one ES on a single node we get major performance degradation due to its requirements. We also recommend host mounts for storage so we would want to ensure we don't have two running on the same node.

> I am more concerned about this part and I need more information so that I am sure it really does what you are trying to do here, so that we dont end up adding something that is not required or does not help.
>
> So when you say any infra pods, what it seems like that basically you are trying to have one such infra pod per node, even though its a soft requirement (preferredDuringSchedulingIgnoredDuringExecution) but unless number of pods is close enough to number of nodes, it wont be of much help here.

Yes, so any other pods that would run on an infra node typically (e.g. router, registry, cassandra/hawkular, service catalog) we would want to balance the distribution across the nodes. I wasn't sure if using `preferredDuringSchedulingIgnoredDuringExecution` would help achieve this or not, perhaps with a different weight?

Reply:
Scheduler balances by default so `preferredDuringSchedulingIgnoredDuringExecution` should help with spreading (not balancing) if more nodes are available. However, if the ratio of number of all infra pods to number of nodes is high, you might not see much benefit. Also in case of node failures, pods from failed nodes would still go to other infra nodes (as `preferredDuringSchedulingIgnoredDuringExecution` a soft requirement) and wont come back automatically to the recovered nodes. 

Please understand, I am not saying you should not do this, and `its ok to go ahead with with these changes`, but understanding how much benefit you would get might help avoid surprises later.

Reply:
Thanks @aveshagarwal ,  this was what I wanted a sanity check on.
@portante do you still feel like this would achieve what you'd like to see regarding other infra pods?

Reply:
So, if we have 3 infra nodes and 3 ES pods, we only want one ES pod per infra node.

If we have 3 infra nodes and 6 ES pods, we only want two ES pods per infra node.

etc.

For multiples it is easy.  But for the case where we have the number of ES pods not as a multiple of ES pods, we just need to be sure that we never have more than `ceiling(# ES pods / # infra nodes)` running on an infra node.

Reply:
@portante then we don't want a `requiredDuringScheduledulingIgnoredDuringExecution` from what you're saying. We should have a preferred for the same component... 
What about in regards to attempting to balance with other infra pods though?

@aveshagarwal is there any way that anti-affinity would be able to handle the scenario that peter described? or is that left up to scheduling in general?

Reply:
> @portante then we don't want a requiredDuringScheduledulingIgnoredDuringExecution from what you're saying.

Yes, it seems that way.

> What about in regards to attempting to balance with other infra pods though?

It seems that Kubernetes does not have the ability to express this kind of scheduling from what I understand, but I could be wrong.

Reply:
IMHO, it would be helpful to discuss what our goals are. As I understand, we are targeting multiple groups of pods in a namespace. Now 2 things we need to consider, 1) requirements within each group, and 2) requirements within all pods in all groups.    

1) requirements within each group: If the goal is `strict` (initial comments by @ewolinetz pointed to this) to have 1 pod per node, then use hard requirement: `requiredDuringScheduledulingIgnoredDuringExecution`. Otherwise (not a strict, means, more than one pod per node are tolerable), try preferredDuringSchedulingIgnoredDuringExecution as within each group it might help somewhat in some scenario like node failures (at least when distribution first pod on each node within a group), but there is no guarantee to have even distribution though scheduler tries its best for even distribution (that is even without preferredDuringSchedulingIgnoredDuringExecution).  

2) requirements within all pods in all groups: This is even trickier, and I presume based on @portante's comments, then when surely dont want  `requiredDuringScheduledulingIgnoredDuringExecution` (as this is not even a requirement within each group, so surely not required for all groups). Now whether `preferredDuringSchedulingIgnoredDuringExecution` would be helpful for this requirement, I doubt that as I have mentioned previously, as soon as the ratio of number of pods to number of nodes is high (which could be the case in this requirement), it might not be much advantageous.

So, in summary, I'd suggest  to go with `preferredDuringSchedulingIgnoredDuringExecution` within each group only.

Reply:
It's possible that we may be crossing multiple namespaces with the pods (as it stands) specifically in regards to the `preferred` rule. The original idea was we could attempt to balance with these other pods.

It seems that the appropriate change for this PR is to remove the current `preferred` rules and change our `required` rules to be `preferred`